recipe-client-addon/lib/StudyStorage.jsm

```diff
@@ -0,0 +1,184 @@
+ *   ISO-formatted date string of when the study was started.
+ * @property {string} studyEndDate
+ *   ISO-formatted date string of when the study was ended.
+ * @property {string} addonId
+ *   Add-on ID for this particular study.
+ * @property {boolean} active
+ *   Is the study still running?
+ */
+
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "IndexedDB", "resource://gre/modules/IndexedDB.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "ajv", "resource://shield-recipe-client/vendor/ajv.js");
+
+this.EXPORTED_SYMBOLS = ["StudyStorage"];
+
+const DB_NAME = "shield-studies";
+const DB_OPTIONS = {
+  version: 1,
+  storage: "persistent",
+};
+const DB_SCHEMA = {
+  type: "object",
+  additionalProperties: false,
+  properties: {
+    name: {
+      type: "string",
+      minLength: 1,
+    },
+    addonId: {
+      type: "string",
+      minLength: 1,
+    },
+    addonVersion: {
+      type: "string",
+      minLength: 1,
+    },
+    description: {
+      type: "string",
+      minLength: 1,
+    },
+    studyStartDate: {
+      type: "string",
+      minLength: 1,
+      format: "date-time",
+    },
+    studyEndDate: {
+      type: ["string", "null"],
+      minLength: 1,
+      format: "date-time",
+      default: null,
+    },
+    active: {
+      type: "boolean",
+      default: true,
+    },
+  },
+};
+const CREATE_SCHEMA = Object.assign({}, DB_SCHEMA, {
+  required: [
+    "name",
+    "addonId",
+    "addonVersion",
+    "description",
+    "studyStartDate",
+  ],
+});
+const UPDATE_SCHEMA = DB_SCHEMA;
+
+const ajvInstance = new ajv({
+  allErrors: true,
+  useDefaults: true,
+});
+const validateCreate = ajvInstance.compile(CREATE_SCHEMA);
+const validateUpdate = ajvInstance.compile(UPDATE_SCHEMA);
+
+/**
+ * Cache the database connection so that it is shared among multiple operations.
+ */
+let databasePromise;
+async function getDatabase() {
+  if (!databasePromise) {
+    databasePromise = IndexedDB.open(DB_NAME, DB_OPTIONS, db => {
+      db.createObjectStore(DB_NAME, {
+        keyPath: "name",
+      });
+    });
+  }
+  return databasePromise;
+}
+
+/**
+ * Get a transaction for interacting with the study store.
+ *
+ * NOTE: Methods on the store returned by this function MUST be called
+ * synchronously, otherwise the transaction with the store will expire.
+ * This is why the helper takes a database as an argument; if we fetched the
+ * database in the helper directly, the helper would be async and the
+ * transaction would expire before methods on the store were called.
+ */
+function getStore(db) {
+  return db.objectStore(DB_NAME, "readwrite");
+}
+
+this.StudyStorage = {
+  async clear() {
+    const db = await getDatabase();
+    await getStore(db).clear();
+  },
+
+  async close() {
+    if (databasePromise) {
+      const db = await databasePromise;
+      await db.close();
+      databasePromise = null;
```

Comment:
Seconded! I noted this in https://github.com/mozilla/normandy/pull/860#discussion_r126279145 and then forgot about it when reviewing this version of the PR. Oops...